keps/sig-node/20190528-node-maintenance-controller.md

```diff
@@ -0,0 +1,190 @@
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Node maintenance controller that performs node drain - server side using custom resources calls to the API in order to perform the drain (cordon + evictions)
+
+## Motivation
+
+Move kubectl drain into the server to enable execution of node drain using custom resources calls to the API in order to perform the drain.
+This would also reduce the proliferation of drain behaviours in other projects that re-implement the drain logic to accustom their needs.
+
+(https://github.com/kubernetes/kubernetes/issues/25625)
+
+### Goals
+
+- A new controller that implements the drain logic that is currently [implemented](https://github.com/kubernetes/kubernetes/blob/v1.14.0/pkg/kubectl/cmd/drain/drain.go#L307) in [kubectl](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#drain).
+- The controller watches and acts upon maintenance CRs to put/remove a node on/from maintenance.
+- Status report of the drain (embodied in the maintenance CR).
+- Custom taints and labels added to the node going under maintenance.
+- Configurable time-out(time buffer) for the drain process.
+
+### Non-Goals
+- "Dry run" mode to detect what is the best node / N nodes to drain
+
+## Proposal
+
+Node maintenance controller watches for new or deleted custom resources e.g. `NodeMaintenance` which indicate that a node in the cluster should either:
```

Comment:
I think one key question is whether this should be implemented as a CRD or as part of the `NodeSpec`.

Reply:
Would this controller be considered a "core" addon? Or an optional addon? Are there any plans to work with sig-cluster-lifecycle to have this controller deployed by default with `kubeadm`?

This is something that we would like to consume within the Cluster API project, and the more integrated it is with the default deployment scenarios the easier it will be for us to adopt.

Reply:
The proposal here aims for core integrated as much as possible, but i guess this could be discussed in later phases. 

I would however add as many use cases and needs for this KEP to strengthen the case, either in this initial PR or in one of the following if it gets merged.